### PR TITLE
Custom Application Sources

### PR DESCRIPTION
**Pull request by derluke**
## Summary
Add support for custom application sources

## Rationale
app sources are faster than building custom docker execution environments for each app, allow for quick change/editting of individual files, and provide support for runtime parameters

### CODE REVIEW

File: src/datarobotx/idp/custom_application_source.py [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+from datarobot.models.runtime_parameters import RuntimeParameterValue
+from datarobot.utils import camelize
+
+from datarobotx.idp.common.hashing import get_hash
+
+
+def get_or_create_custom_application_source(endpoint: str, token: str, name: str) -> str:
+    """Get or create a custom application source with the given name.
+
+    Args:
```
Comment (mpkrass7) [inline src/datarobotx/idp/custom_application_source.py:35]: Small nit: in drx we followed [numpy docstring](https://numpydoc.readthedocs.io/en/latest/format.html) format. I think we're also following that here. Also in most of our helpers we did not explicitly write out endpoint and token in the docstring as arguments though I don't think we've ever explicitly said to exclude them.
https://github.com/datarobot-community/datarobotx-idp/blob/main/src/datarobotx/idp/common/handle_io.py#L101

File: src/datarobotx/idp/custom_application_source.py [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+    custom_application_source_id = create_custom_application_source_response.json()["id"]
+
+    client.patch(
+        f"customApplicationSources/{custom_application_source_id}",
+        json={"name": name},
+    )
+    return str(custom_application_source_id)
+
+
+# Helper functions
```
Comment (mpkrass7) [inline src/datarobotx/idp/custom_application_source.py:67]: _leading underscore signifies helper I think so suggest removing comment

File: src/datarobotx/idp/custom_application_source.py [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+def _find_existing_custom_application_source_version(
+    endpoint: str, token: str, custom_application_source_id: str, version_token: str
+) -> str:
+    client = dr.Client(endpoint=endpoint, token=token)  # type: ignore
+    versions = client.get(
+        f"customApplicationSources/{custom_application_source_id}/versions/"
+    ).json()["data"]
+    for version in versions:
+        if version.get("label") and version_token in version["label"]:
```
Comment (mpkrass7) [inline src/datarobotx/idp/custom_application_source.py:76]: No need to change but if you'd like this line to be shorter you could do:
`if version_token in version.get("label", "")`

File: src/datarobotx/idp/custom_application_source.py [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+                                for param in runtime_parameter_values_objs
+                            ]
+                        ),
+                    )
+                )
+
+            if "replicas" in kwargs:
+                upload_data.append(("replicas", str(kwargs["replicas"])))
+
+            if base_environment_id := kwargs.get("base_environment_id"):
```
Comment (mpkrass7) [inline src/datarobotx/idp/custom_application_source.py:154]: I keep forgetting to use the walrus operator but I love seeing it every time

File: src/datarobotx/idp/custom_application_source.py [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+    previous_custom_application_source_version_id : str
+    runtime_parameter_values : Optional[List[Union[RuntimeParameterValue, Dict[str, str]]]]
+    kwargs : Any
+
+    Returns
+    -------
+    str
+        The ID of the custom application source version.
+    """
+    return _get_or_create_custom_application_source_version(
```
Comment (mpkrass7) [inline src/datarobotx/idp/custom_application_source.py:249]: Ok I get the idea now. Small nit but consider doing: 
1. Only one public function `get_or_create_custom_application_source_version`
2. `previous_custom_application_source_version_id` is an optional argument which defaults to None
3.  If a `previous_custom_application_source_version_id` is supplied, we assume a patch request in the helper. Otherwise post request
4. `from_previous` is removed from helper function

File: src/datarobotx/idp/custom_application_source.py [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+        )
+
+    try:
+        existing_version_id = _find_existing_custom_application_source_version(
+            endpoint=endpoint,
+            token=token,
+            custom_application_source_id=custom_application_source_id,
+            version_token=version_token,
+        )
+        return existing_version_id
```
Comment (mpkrass7) [inline src/datarobotx/idp/custom_application_source.py:126]: Actually reading this again, I think there is a logical problem: 

Lets say user passes `from_previous` to this function. But their runtime parameters are different. You create a token from the version already on DataRobot but that token doesn't involve the runtime parameters. When you try to return the `existing_version_id`, it succeeds and the runtime parameters are therefore never updated

File: src/datarobotx/idp/custom_application_source.py (hunk 5 above)
Comment (derluke) [reply src/datarobotx/idp/custom_application_source.py:249]: I had that, but then decided to follow the custom_model_version API here. Where dr SDK also has two methods, and we have two different functions

File: src/datarobotx/idp/custom_application_source.py (hunk 6 above)
Comment (derluke) [reply src/datarobotx/idp/custom_application_source.py:126]: right. agree

File: src/datarobotx/idp/custom_applications.py [hunk 7]
```diff
@@ -68,24 +71,46 @@ def _list_custom_apps(endpoint: str, token: str, name: Optional[str] = None) ->
...
+    env_version_id: Optional[str] = None,
+    custom_application_source_version_id: Optional[str] = None,
     **kwargs: Any,
 ) -> str:
     """Get, replace, or create a custom application from a custom environment with requested parameters.
 
     If a custom app with the desired name already exists but with different parameters, the existing
     app will be deleted and a new one created.
+
+    Exactly one of env_version_id or custom_application_source_version_id is required.
```
[the remark this answers is not in the document]
Comment (derluke) [reply src/datarobotx/idp/custom_applications.py:93]: oh, looking at the reply from client.get("customApplications/"):
neither environmentId or baseEnvironmentId are set. only envVersionId xor customApplicationSource and customApplicationSourceVersionId

But envVersionId is not allowed to create an app.. I think I understand.

File: src/datarobotx/idp/custom_application_source.py (hunk 5 above)
Comment (derluke) [reply src/datarobotx/idp/custom_application_source.py:249]: I just found out that from_previous works differently to custom models. Custom models you get a new version with the content freom before. Here you get the same version with new content (rather interestingly..)

I think the function needs to be renamed

Comment (derluke) [reply src/datarobotx/idp/custom_application_source.py:249]: Or maybe the function should be removed from idp?

Comment (derluke) [reply src/datarobotx/idp/custom_application_source.py:249]: yeah. thank you for coming to my ted talk. I agree with you

File: src/datarobotx/idp/custom_applications.py [hunk 8]
```diff
@@ -68,24 +71,58 @@ def _list_custom_apps(endpoint: str, token: str, name: Optional[str] = None) ->
 
 def _find_existing_custom_app(endpoint: str, token: str, **kwargs: Any) -> str:
     for app in _list_custom_apps(endpoint, token):
-        if all([app[camelize(key)] == kwargs[key] for key in kwargs]):
+        if all([app[camelize(key)] == kwargs[key] for key in kwargs if kwargs[key] is not None]):
```
Comment (brau0300) [inline src/datarobotx/idp/custom_applications.py:74]: why is this change necessary? what i don't like about it is it opens up the possibility that user explicitly requested None for something but the server has something other than None for that same thing

File: src/datarobotx/idp/custom_applications.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (brau0300) [reply src/datarobotx/idp/custom_applications.py:94]: So is "Exactly one of env_version_id or custom_application_source_version_id is required." correct? under what conditions is "environment_id" required?

File: src/datarobotx/idp/custom_application_source.py [hunk 9]
```diff
@@ -0,0 +1,316 @@
+#
```
Comment (brau0300) [inline src/datarobotx/idp/custom_application_source.py:1]: on the surface - feels like this should be two files to maintain precedent/consistency, e.g. custom_application_source.py, custom_application_source_version.py; i also see this is how you organized your tests

File: src/datarobotx/idp/custom_application_source.py [hunk 10]
```diff
@@ -0,0 +1,316 @@
...
+        return str(new_version)
+
+
+# Public functions
+def get_or_create_custom_application_source_version(
+    endpoint: str,
+    token: str,
+    custom_application_source_id: str,
+    runtime_parameter_values: Optional[List[Union[RuntimeParameterValue, Dict[str, str]]]] = None,
+    previous_custom_application_source_version_id: Optional[str] = None,
```
Comment (brau0300) [inline src/datarobotx/idp/custom_application_source.py:184]: this parameter is not present in the rest API (is idiosyncratic to idp) and adds a lot complexity (risk surface for mistakes); i don't think we have this capability for custom model version - do we really need it here?

File: src/datarobotx/idp/custom_application_source.py [hunk 11]
```diff
@@ -0,0 +1,316 @@
...
+        endpoint=endpoint,
+        token=token,
+        custom_application_source_id=custom_application_source_id,
+        runtime_parameter_values=runtime_parameter_values,
+        previous_custom_application_source_version_id=previous_custom_application_source_version_id,
+        **kwargs,
+    )
+
+
+# def get_or_create_custom_application_source_version_from_previous(
```
Comment (brau0300) [inline src/datarobotx/idp/custom_application_source.py:217]: can be deleted?

File: src/datarobotx/idp/custom_application_source.py [hunk 12]
```diff
@@ -0,0 +1,316 @@
...
+#         token=token,
+#         custom_application_source_id=custom_application_source_id,
+#         previous_custom_application_source_version_id=previous_custom_application_source_version_id,
+#         runtime_parameter_values=runtime_parameter_values,
+#         **kwargs,
+#     )
+
+
+def get_or_create_qanda_app(
+    endpoint: Optional[str],
```
Comment (brau0300) [inline src/datarobotx/idp/custom_application_source.py:257]: we are moving away from altering the signature for endpoint/token unless we standardize everywhere, no?

File: src/datarobotx/idp/custom_application_source.py [hunk 13]
```diff
@@ -0,0 +1,316 @@
...
+        The name of the Q&A app.
+
+    Returns
+    -------
+    str
+        The ID of the Q&A app.
+    """
+    app_token = get_hash(name, deployment_id, environment_id)
+
+    if endpoint is not None and token is not None:
```
Comment (brau0300) [inline src/datarobotx/idp/custom_application_source.py:285]: if aligned on above comment re: matching signature this goes away

File: tests/integration/test_custom_application_source_version.py [hunk 14]
```diff
@@ -0,0 +1,231 @@
...
+        ).json()["items"]
+    ]
+    env_ver_id_4 = get_or_create_custom_application_source_version(
+        endpoint=dr_endpoint,
+        token=dr_token,
+        custom_application_source_id=custom_application_source,
+        previous_custom_application_source_version_id=env_ver_id_3,
+        folder_path=additional_metadata_context_path,
+    )
+    assert env_ver_id_3 == env_ver_id_4
```
Comment (brau0300) [inline tests/integration/test_custom_application_source_version.py:184]: see my comment above - this behavior feels idiosyncratic and doesn't align to our precedent in custom model versions

Comment (brau0300) [reply tests/integration/test_custom_application_source_version.py:184]: ditto to the below assertions on the id staying the same

File: src/datarobotx/idp/custom_application_source.py [hunk 15]
```diff
@@ -0,0 +1,316 @@
...
+                                for param in runtime_parameter_values_objs
+                            ]
+                        ),
+                    )
+                )
+
+            if "replicas" in kwargs:
+                upload_data.append(("replicas", str(kwargs["replicas"])))
+
+            if base_environment_id := kwargs.get("base_environment_id"):
```
Comment (brau0300) [inline src/datarobotx/idp/custom_application_source.py:144]: i think this is actually a different conditional; you want to test if "base_environment_id" is in kwargs which is not strictly equivalent to whether get returns a none-Falsey value; better to be explicit IMO; same to the below test on base env version id

File: src/datarobotx/idp/custom_applications.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (derluke) [reply src/datarobotx/idp/custom_applications.py:94]: env_version_id is not required, and not even supported to create an app - but is then set behind the scenes to the app (I guess using the latest version?)

one and only one of environment version or custom app source or custom app source version is required.

File: src/datarobotx/idp/custom_applications.py (hunk 8 above)
Comment (derluke) [reply src/datarobotx/idp/custom_applications.py:74]: The issue is that env_version_id is being set automatically. If provided or not. And that then fails the test if it wasn't set in the call to get_replace_or_create_custom_app

File: src/datarobotx/idp/custom_application_source.py (hunk 13 above)
Comment (derluke) [reply src/datarobotx/idp/custom_application_source.py:285]: We need the from_previous - otherwise the environment is wiped

File: src/datarobotx/idp/custom_application_source.py (hunk 15 above)
Comment (derluke) [reply src/datarobotx/idp/custom_application_source.py:144]: I don't understand the concern

Comment (brau0300) [reply src/datarobotx/idp/custom_application_source.py:144]:
```
if "base_environment_id" in kwargs:
    upload_data.append(("baseEnvironmentId", kwargs['base_environment_id'))
```
Is both more clear/readable on first glance and also doesn't cause the reader/reviewer to start thinking about corner cases and whether they apply/need to be managed (e.g. what if user explicitly passes in None or a falsey value for this kwarg); the above basically says if user explicitly passed in base_environment_id, pass through what they requested; the original has the walrus operator complication and also makes the reader think about how does the get() work in concert with the walrus if falsey values are explicitly passed in? what if nothing was explicitly passed in? etc.

File: src/datarobotx/idp/custom_applications.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (brau0300) [reply src/datarobotx/idp/custom_applications.py:94]: i'm still confused after re-reading this thread and the docstring; can we make the docstring more clear (i am assuming the function signature is correct and all three of these are indeed Optional)

File: tests/integration/test_custom_application_source_version.py [hunk 16]
```diff
@@ -174,32 +175,32 @@ def test_get_or_create(
             f"customApplicationSources/{custom_application_source}/versions/{env_ver_id_3}"
         ).json()["items"]
     ]
-    env_ver_id_4 = get_or_create_custom_application_source_version(
+    env_ver_id_4 = get_or_create_custom_application_source_version_from_previous(
         endpoint=dr_endpoint,
         token=dr_token,
         custom_application_source_id=custom_application_source,
-        previous_custom_application_source_version_id=env_ver_id_3,
+        # previous_custom_application_source_version_id=env_ver_id_3,
```
Comment (brau0300) [inline tests/integration/test_custom_application_source_version.py:182]: assuming these can be deleted?

File: tests/integration/test_custom_application_source_version.py [hunk 17]
```diff
@@ -174,32 +175,32 @@ def test_get_or_create(
             f"customApplicationSources/{custom_application_source}/versions/{env_ver_id_3}"
         ).json()["items"]
     ]
-    env_ver_id_4 = get_or_create_custom_application_source_version(
+    env_ver_id_4 = get_or_create_custom_application_source_version_from_previous(
```
Comment (brau0300) [inline tests/integration/test_custom_application_source_version.py:178]: i still am not tracking why we can't create this same experience (e.g. test will pass) using a single `get_or_create_custom_application_source_version` ?

**Review by brau0300 — CHANGES_REQUESTED** [review]

File: tests/integration/test_custom_application_source_version.py (hunk 17 above)
Comment (derluke) [reply tests/integration/test_custom_application_source_version.py:178]: Somehow I have to differentiate between creating a new thing from scratch vs building on top of an existing thing

Comment (derluke) [reply tests/integration/test_custom_application_source_version.py:178]: Surely we have to have something that reflects this:

image

Comment (derluke) [reply tests/integration/test_custom_application_source_version.py:178]: which is the same as this (in custom model workshop)
image

Comment (derluke) [reply tests/integration/test_custom_application_source_version.py:178]: the change I introduced is that none of our functions allow for modifying an existing version in place. it's always a new version, or retrieve existing. No more version content changes but version id stays the same

File: src/datarobotx/idp/custom_applications.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (derluke) [reply src/datarobotx/idp/custom_applications.py:94]: I changed the doc string to correctly say "environment_id".

they are not optional, at least one should be provided. We can also add custom_application_source_id on top of the version - but it wasn't required so far

File: src/datarobotx/idp/custom_application_source_version.py [hunk 18]
```diff
@@ -0,0 +1,256 @@
...
+    """
+    if runtime_parameter_values is not None:
+        runtime_parameter_values_objs = [
+            RuntimeParameterValue(**d) for d in runtime_parameter_values if isinstance(d, dict)
+        ]
+    else:
+        runtime_parameter_values_objs = None
+
+    client = dr.Client(token=token, endpoint=endpoint)  # type: ignore
+    app_version_token = get_hash(
```
Comment (brau0300) [inline src/datarobotx/idp/custom_application_source_version.py:218]: i think the previous verison id lookup needs to happen before this and the id should be included in the hash for this to be idempotent; also we should raise an error if this lookup fails (e.g. no previous versions at all; this is organized a bit different from the CMV helper and i haven't thought deeply about if that's concerning, but am just noting it now for discussion as needed

**Review by brau0300 — CHANGES_REQUESTED** [review]

File: src/datarobotx/idp/custom_application_source_version.py [hunk 19]
```diff
@@ -0,0 +1,256 @@
...
+            new_version = response.json()["id"]
+
+        return str(new_version)
+
+
+# Public functions
+def get_or_create_custom_application_source_version(
+    endpoint: str,
+    token: str,
+    custom_application_source_id: str,
```
[the remark this answers is not in the document]
Comment (derluke) [reply src/datarobotx/idp/custom_application_source_version.py:148]:
```
from datarobotx.idp.custom_application_source import get_or_create_custom_application_source
from datarobotx.idp.custom_application_source_version import get_or_create_custom_application_source_version

custom_application_source = get_or_create_custom_application_source(endpoint=endpoint, token=token, name="test")
custom_application_source_version = get_or_create_custom_application_source_version(endpoint=endpoint, token=token, custom_application_source_id=custom_application_source)
```
works. It's not immediately useful (base_environment is empty, so it won't work - but it creates without errors)

[the remark this answers is not in the document]
Comment (derluke) [reply src/datarobotx/idp/custom_application_source_version.py:148]: I did check with the team - they said "We do not plan to change the APIs but that is not guaranteed before GA"

File: src/datarobotx/idp/custom_application_source_version.py (hunk 18 above)
Comment (derluke) [reply src/datarobotx/idp/custom_application_source_version.py:218]:
```
except:
    raise ValueError("No previous version found")
```
in line 254 is raising the error if there are no previous versions

I am using the hash in the call to find the previous version - so I need to calculate the token before, no?

The whole thing got slightly confusing with several refactors. I was relying on the tests to convince myself of the correct behaviour

Comment (derluke) [reply src/datarobotx/idp/custom_application_source_version.py:242]: I refactored out one function to make it a bit more clear what's happening.

**Review by brau0300 — APPROVED** [review]
